test/langtools/tools/javac/sealed/PrivateMembersInPermitClause.java

```diff
@@ -0,0 +1,158 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.nio.file.Path;
+
+/**
+ * @test
+ * @bug 8338981
+ * @summary Access to private classes should be permitted inside the permits clause of the enclosing top-level class
+ * @library /tools/lib
+ * @modules jdk.compiler/com.sun.tools.javac.api
+ * jdk.compiler/com.sun.tools.javac.main
+ * jdk.compiler/com.sun.tools.javac.util
+ * @build toolbox.ToolBox toolbox.JavacTask toolbox.Task
+ * @run main PrivateMembersInPermitClause
+ */
+
+public class PrivateMembersInPermitClause extends toolbox.TestRunner {
+
+    private final toolbox.ToolBox tb;
+
+    public PrivateMembersInPermitClause() {
+        super(System.err);
+        tb = new toolbox.ToolBox();
+    }
+
+    public static void main(String... args) throws Exception {
+        new PrivateMembersInPermitClause().runTests();
+    }
+
+    public void runTests() throws Exception {
+        runTests(_ -> new Object[] {});
+    }
+
+    @Test
+    public void givenPrivateClassInPermitsClause_whenCompiling_thenShouldCompile() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            sealed class S permits S.A {
+                private static final class A extends S {}
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"))
+            .run(toolbox.Task.Expect.SUCCESS);
+    }
+
+    @Test
+    public void givenPrivateClassOfOtherTopLevelClassInPermitsClause_whenCompiling_thenShouldFail() throws Exception {
+        var root = Path.of("src");
+        tb.writeJavaFiles(root,
+            """
+            public class S {
+                private static final class A extends S {}
+            }
+            """
+        );
+
+        tb.writeJavaFiles(root,
+            """
+            public sealed class T permits S.A {
+                private static final class A extends T {}
+            }
+            """
+        );
+
+        new toolbox.JavacTask(tb)
+            .files(root.resolve("S.java"), root.resolve("T.java"))
+            .run(toolbox.Task.Expect.FAIL);
```

Comment:
Negative tests should, unless there is a very good reason not to, verify the exact errors printed. Use `-XDrawDiagnostics` option, and `.getOutputLines(Task.OutputKind.DIRECT)` on the result to get the output lines. (And, preferably also `.writeAll()` on all `toolbox.JavacTask` runs, passing or failing.) See for example here:

https://github.com/openjdk/jdk/blob/855c8a7def21025bc2fc47594f7285a55924c213/test/langtools/tools/javac/ImportModule.java#L175